results/pkg/watcher/reconciler/taskrun/taskrun.go

```diff
@@ -0,0 +1,99 @@
+package taskrun
+
+import (
+	"context"
+
+	"github.com/tektoncd/experimental/results/pkg/watcher/convert"
+	"github.com/tektoncd/experimental/results/pkg/watcher/reconciler/common"
+	pb "github.com/tektoncd/experimental/results/proto/proto"
+	"github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	listers "github.com/tektoncd/pipeline/pkg/client/listers/pipeline/v1beta1"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/logging"
+)
+
+type Reconciler struct {
+	logger            *zap.SugaredLogger
+	client            pb.ResultsClient
+	taskRunLister     listers.TaskRunLister
+	pipelineclientset versioned.Interface
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Errorf("Invalid resource key: %s", key)
+		return nil
+	}
+	logger.With(zap.String("Namespace", namespace), zap.String("Name", name))
+
+	tr, err := r.pipelineclientset.TektonV1beta1().TaskRuns(namespace).Get(name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		logger.Errorf("TaskRun in work queue no longer exists: %v", err)
+		return nil
+	}
+	if err != nil {
+		logger.Errorf("Error retrieving TaskRun: %v", err)
+		return err
+	}
+
+	logger.Info("Recieving new TaskRun")
+	trProto, err := convert.ToTaskRunProto(tr)
+	if err != nil {
+		logger.Errorf("Error converting TaskRun to its corresponding proto: %v", err)
+		return err
+	}
+
+	if resultID, ok := trProto.GetMetadata().GetAnnotations()[common.IDName]; ok {
```

Comment:
This looks highly duplicated with pipelinerun, when all that differs seems to be the Execution. We can abstract and combine the duplicative logic?

Reply:
Yeah,, thanks for bring this out! I've tried to pull this logic out as a common function, however, when it comes to the Execution, I need to specify the type, just like `execution.GetPipelineRun()` and `&pb.Execution{Execution: &pb.Execution_PipelineRun{prProto}}`, and they appear a lot, so pulling the logic out as a common function requires a lot of type switches. 
Moreover, the logic seems like the same in the current stage, but in the future, we may want to make different changes for different type. Therefore I prefer we just keep them sparated.